### ISSUE TITLE
What to do when: tip of your current branch is behind hint its remote counterpart?

### ISSUE THREAD

**Issue by saratrem** [STAT545-UBC/Discussion#364]
Hi, 

I created a sub-directory with Git (not with R studio) but now when I tried to push my homework 4 to my github repertory I get this error message: 

To https://github.com/STAT545-UBC/tremblay-boyer_sara.git
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/STAT545-UBC/tremblay-boyer_sara.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details

So I tried to follow this help article " Dealing with non-fast forward errors"  https://help.github.com/articles/dealing-with-non-fast-forward-errors/ and also thishttps://help.github.com/articles/pushing-to-a-remote/

but it didn't really help,  I'm still stuck!  

Thank you 

Sara

**Comment by jennybc** [comment]
First, some of the above doesn't quite make sense. I'm not trying to be difficult -- it's just really hard to answer these questions remotely, regardless. And especially hard when things don't add up.

> I created a sub-directory with Git (not with R studio) 

Creating a sub-directory is a file system operation. It means "I created a folder on my computer". Git cannot do this. So I'm not sure what you mean here. Did you create a subdirectory? Or did you, possibly by mistake, create a Git branch, which is something entirely different?


But let's set that aside. The above description makes it sound like there are changes on GitHub that you have not pulled to your local computer. It sounds like you need to pull. And possibly resolve conflicts? I somehow suspect that will be the case 😥. Once that's done, you should be able to push.

**Comment by jennybc** [comment]
The other explanation for your situation is that you somehow went backwards in time on your local computer, i.e. reset to an earlier commit or undid a commit. But those commits had already been pushed to GitHub. Which results in the same thing: Git thinks that your repo on GitHub is "ahead" of your local repo.

**Comment by saratrem** [comment]
Well I created a new folder (named: Hw04) with one file inside (named README.md) in my repository ... I thought that was a sub-directory?!  I dont think I created a branch. 

I did try to pull earlier and I got this message: 

error: Pull is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.

I dont really know what " work tree" refers to but  i tried from my shell to add/ rm the README.md file I just created but it does not work.   

Thank you

**Comment by oganm** [comment]
Did you commit your latest local changes before attempting to pull? As long as you have uncommitted changes you cannot pull.

As a side note:
Your probably want to solve this without doing what I'm about to say just to learn what went wrong but as a surefire way to resolve all conflicts, you can just 

- rename your repo locally
- re-clone it from github
-  paste the changes you want to keep from the renamed repo
- delete the renamed repo
- commit and push

You just have to be careful not to delete anything you want to keep by accident and don't forget to feel some shame every time you do this.

**Comment by jennybc** [comment]
> Well I created a new folder (named: Hw04) with one file inside (named README.md) in my repository ... I thought that was a sub-directory?! I dont think I created a branch.

Yes that is a sub-directory! Great! In your first message, **you said you created it with Git**, which is impossible. That is why I was worried that maybe you meant you created a branch. But sounds like we don't need to worry about this any more.

> don't forget to feel some shame every time you do this.

Oh I disagree! About the shame part. When you're new-ish, this can be a very safe and effective way to undo whatever git mess you've made. Especially if it's not convenient to come to office hours or get hands on help. So, minus the shame, I think @oganm's suggestion is an excellent way for you to proceed @saratrem.

At some point, you obviously got into a Git mess. And that went unresolved. Now, though your local files may be exactly the way you want, your Git infrastructure is screwed up. And it's keeping you from moving forward.

Highly you recommend you do as @oganm says.

**Comment by saratrem** [comment]
I commited everything even old random things I didn't want to have in my repo and tried to push again and for some unknown reason, it worked! So didnt even have to feel ashamed to rename/delete my repo. My repo is pretty messy, but I will go to office hours to fix it. Thank you for your time!
